packages/ddp-client/livedata_connection.js

```diff
@@ -873,42 +821,255 @@ _.extend(Connection.prototype, {
       // block, and add it to the last block of methods.
-      if (_.isEmpty(self._outstandingMethodBlocks) ||
-          _.last(self._outstandingMethodBlocks).wait)
-        self._outstandingMethodBlocks.push({wait: false, methods: []});
-      _.last(self._outstandingMethodBlocks).methods.push(methodInvoker);
+      let lastBlock = _.last(this._outstandingMethodBlocks);
+      if (! lastBlock || lastBlock.wait) {
+        lastBlock = { wait: false, methods: [] };
+        this._outstandingMethodBlocks.push(lastBlock);
+      }
+      lastBlock.methods.push(methodInvoker);
     }
 
     // If we added it to the first block, send it out now.
-    if (self._outstandingMethodBlocks.length === 1)
+    if (this._outstandingMethodBlocks.length === 1) {
       methodInvoker.sendMessage();
+    }
+  },
 
-    // If we're using the default callback on the server,
-    // block waiting for the result.
-    if (future) {
-      return future.wait();
+  callAsync(name, ...args) {
+    return this.applyAsync(name, args);
+  },
+
+  applyAsync(name, args, options = {}) {
+    // Keep our args safe from mutation (eg if we don't send the message for a
+    // while because of a wait method).
+    args = EJSON.clone(args);
+
+    const methodId = this._makeLazyMethodIdGetter();
+    const enclosing = DDP._CurrentInvocation.get();
+    const alreadyInSimulation = enclosing && enclosing.isSimulation;
+    const randomSeedGenerator =
+      this._makeRandomSeedGenerator(enclosing, name);
+
+    const stubResult = this._simulateMethod({
+      name, args, methodId, randomSeedGenerator
+    });
+
+    const stubPromise = Promise.resolve(stubResult && stubResult.promise);
+
+    // If we're in a simulation, stop and return the result we have,
+    // rather than going on to do an RPC. If there was no stub, the value
+    // of afterStubPromise will end up being undefined.
+    if (alreadyInSimulation) {
+      return stubPromise;
+    }
+
+    // At this point we're definitely doing an RPC, and we're going to
+    // return the value of the RPC to the caller.
+
+    let rpcPromise;
+    const ensureRPCSent = () => {
+      if (rpcPromise) {
+        return rpcPromise;
+      }
+
+      // Send the RPC. Note that on the client, it is important that the
+      // stub have finished before we send the RPC, so that we know we have
+      // a complete list of which local documents the stub wrote.
+      const message = {
+        msg: 'method',
+        method: name,
+        params: args,
+        id: methodId()
+      };
+
+      // Send the randomSeed only if we used it
+      if (randomSeedGenerator.seed) {
+        message.randomSeed = randomSeedGenerator.seed;
+      }
+
+      return rpcPromise = new Promise((resolve, reject) => {
+        this._enqueueMethodInvoker(new MethodInvoker({
+          methodId: methodId(),
+          callback: (error, result) => {
+            error ? reject(error) : resolve(result);
+          },
```

Comment:
Can do the full shorthand for an object literal function instead of arrow syntax

```js
callback(error, result) {
  error ? reject(error) : resolve(result);
}
```